src/mlpack/core/tree/vantage_point_tree/vantage_point_split.hpp

```diff
@@ -0,0 +1,197 @@
+   * @param oldFromNew Vector which will be filled with the old positions for
+   *    each new point.
+   */
+  static bool SplitNode(const BoundType& bound,
+                        MatType& data,
+                        const size_t begin,
+                        const size_t count,
+                        size_t& splitCol,
+                        std::vector<size_t>& oldFromNew);
+ private:
+  /**
+   * The maximum number of samples used for vantage point estimation and for
+   * estimation of the median.
+   */
+  static const size_t maxNumSamples = 100;
+
+  template<typename StructElemType>
+  struct SortStruct
+  {
+    size_t point;
+    ElemType dist;
+  };
+
+  template<typename StructElemType>
+  static bool StructComp(const SortStruct<StructElemType>& s1,
+      const SortStruct<StructElemType>& s2)
+  {
+    return (s1.dist < s2.dist);
+  };
+
+  /**
+   * Select the best vantage point i.e. the point with the largest second moment
+   * of the distance from a number of random node points to the vantage point.
+   * Firstly this methods selects no more than maxNumSamples random points.
+   * Then it evaluates each point i.e. calcilates the corresponding second
+   * moment and selects the point with the largest moment. Each random point
+   * belongs to the node.
+   *
+   * @param bound The bound used for this node.
+   * @param data The dataset used by the binary space tree.
+   * @param begin Index of the starting point in the dataset that belongs to
+   *    this node.
+   * @param count Number of points in this node.
+   * @param vantagePoint The index of the vantage point in the dataset.
+   * @param mu The median value of distance form the vantage point to
+   * a number of random points.
+   */
+  static void SelectVantagePoint(const BoundType& bound, const MatType& data,
+    const size_t begin, const size_t count, size_t& vantagePoint, ElemType& mu);
+
+  /**
+   * Find no more then max(numSamples, upperBound) random samples i.e.
+   * random points that belong to the node. Each sample belongs to
+   * the interval [begin, begin + upperBound)
+   *
+   * @param distinctSamples The vector of samples indices.
+   * @param numSamples Maximum number of samples.
+   * @param begin The least index.
+   * @param upperBound The upper bound of indices.
+   */
+  static void GetDistinctSamples(arma::uvec& distinctSamples,
+      const size_t numSamples, const size_t begin, const size_t upperBound);
+
+  /**
+   * Get the median value of the distance from a certain vantage point to a
+   * number of samples.
+   *
+   * @param bound The bound used for this node.
+   * @param data The dataset used by the binary space tree.
+   * @param samples The indices of random samples.
+   * @param vantagePoint The vantage point.
+   * @param mu The median value.
+   */
+  static void GetMedian(const BoundType& bound, const MatType& data,
+      const arma::uvec& samples,  const size_t vantagePoint, ElemType& mu);
```

Comment:
I think here and in `GetSecondMoment()`, it would be better to pass a `MetricType` instead of a `BoundType`.